# The data layer: the frozen panel and its coverage

*A learning exercise performed in role: a simulated mandate with no client and no institution. Nothing in this notebook is investment advice, a recommendation, or a client communication.*

**Entry point** `python3 -m portfolio_workbench.data.coverage`

**Modules covered** `data/coverage.py`, `data/loader.py`, `data/manifest.py`, `data/panel.py`, `data/quality.py`, `data/universe.py`, `data/external.py`

_Generated from the code by `python3 -m reporting.notebooks`: the module headers below are read out of the modules themselves, and the run is the entry point's own output._

## 1. What this module does, and the source of every method in it

The data layer turns three external sources into one monthly table: a frozen price snapshot of eleven UCITS sleeves, the European Central Bank's overnight rate, and the Fama/French factor archives. Every other layer reads that table and nothing else. The coverage entry point prints what arrived, what was dropped, and from which month each sleeve is priced.

**Sources.** Every public function of the modules this notebook covers, and what it traces to. The map is checked over the code by the acceptance fixture, so a method added without a source fails a command rather than going unnoticed.

- `data/coverage.py::main` traces to the coverage report of this effort: the joined panel's shape, its spans and the sleeves' own coverage, printed rather than summarised so a gap is visible before it is used
- `data/loader.py::load_factors` traces to the snapshot contract of this effort: a manifest plus one long-format file per leg, read as monthly bars keyed to the instrument and the period
- `data/loader.py::load_panel` traces to the snapshot contract of this effort: a manifest plus one long-format file per leg, read as monthly bars keyed to the instrument and the period
- `data/loader.py::load_risk_free` traces to the snapshot contract of this effort: a manifest plus one long-format file per leg, read as monthly bars keyed to the instrument and the period
- `data/loader.py::snapshot_root` traces to the snapshot contract of this effort: a manifest plus one long-format file per leg, read as monthly bars keyed to the instrument and the period
- `data/manifest.py::build` traces to the manifest of this effort: what makes two runs the same run is the snapshot they read, so the instrument facts, the file hashes and the vintages are recorded rather than inferred
- `data/manifest.py::describe` traces to the manifest of this effort: what makes two runs the same run is the snapshot they read, so the instrument facts, the file hashes and the vintages are recorded rather than inferred
- `data/manifest.py::measure` traces to the manifest of this effort: what makes two runs the same run is the snapshot they read, so the instrument facts, the file hashes and the vintages are recorded rather than inferred
- `data/manifest.py::read` traces to the manifest of this effort: what makes two runs the same run is the snapshot they read, so the instrument facts, the file hashes and the vintages are recorded rather than inferred
- `data/manifest.py::sha256` traces to the manifest of this effort: what makes two runs the same run is the snapshot they read, so the instrument facts, the file hashes and the vintages are recorded rather than inferred
- `data/manifest.py::verify` traces to the manifest of this effort: what makes two runs the same run is the snapshot they read, so the instrument facts, the file hashes and the vintages are recorded rather than inferred
- `data/panel.py::add_availability` traces to the table contract of this effort: one row per instrument-month carrying `period_month` and `available_from`, the as-of rule that gates every join, and the total-return proxy of the adjusted close
- `data/panel.py::as_of` traces to the table contract of this effort: one row per instrument-month carrying `period_month` and `available_from`, the as-of rule that gates every join, and the total-return proxy of the adjusted close
- `data/panel.py::available_from` traces to the as-of rule of this effort: a month's bar is available only from the next month's first day, which is what makes the join look-ahead-free
- `data/panel.py::available_months` traces to the table contract of this effort: one row per instrument-month carrying `period_month` and `available_from`, the as-of rule that gates every join, and the total-return proxy of the adjusted close
- `data/panel.py::coverage_months` traces to the table contract of this effort: one row per instrument-month carrying `period_month` and `available_from`, the as-of rule that gates every join, and the total-return proxy of the adjusted close
- `data/panel.py::currency_split` traces to the table contract of this effort: one row per instrument-month carrying `period_month` and `available_from`, the as-of rule that gates every join, and the total-return proxy of the adjusted close
- `data/panel.py::cutoff` traces to the table contract of this effort: one row per instrument-month carrying `period_month` and `available_from`, the as-of rule that gates every join, and the total-return proxy of the adjusted close
- `data/panel.py::eur_excess_returns` traces to the panel's own series: euro total returns translated from the local leg, less the accrued cash rate
- `data/panel.py::excess` traces to the table contract of this effort: one row per instrument-month carrying `period_month` and `available_from`, the as-of rule that gates every join, and the total-return proxy of the adjusted close
- `data/panel.py::joined_months` traces to the table contract of this effort: one row per instrument-month carrying `period_month` and `available_from`, the as-of rule that gates every join, and the total-return proxy of the adjusted close
- `data/panel.py::month_span` traces to the table contract of this effort: one row per instrument-month carrying `period_month` and `available_from`, the as-of rule that gates every join, and the total-return proxy of the adjusted close
- `data/panel.py::recomputed_total_return` traces to the table contract of this effort: one row per instrument-month carrying `period_month` and `available_from`, the as-of rule that gates every join, and the total-return proxy of the adjusted close
- `data/panel.py::sleeve_order` traces to the table contract of this effort: one row per instrument-month carrying `period_month` and `available_from`, the as-of rule that gates every join, and the total-return proxy of the adjusted close
- `data/panel.py::total_return` traces to the total-return proxy of this effort: the adjusted close, with a distribution-blind series refused rather than patched
- `data/panel.py::wide` traces to the table contract of this effort: one row per instrument-month carrying `period_month` and `available_from`, the as-of rule that gates every join, and the total-return proxy of the adjusted close
- `data/panel.py::window` traces to the table contract of this effort: one row per instrument-month carrying `period_month` and `available_from`, the as-of rule that gates every join, and the total-return proxy of the adjusted close
- `data/quality.py::fund_size_eur` traces to the eight stops and four warnings of this effort, each taken from the European UCITS universe audit's findings, plus the issuer facts the UCITS rules make checkable
- `data/quality.py::income_policy` traces to the eight stops and four warnings of this effort, each taken from the European UCITS universe audit's findings, plus the issuer facts the UCITS rules make checkable
- `data/quality.py::issuer_facts` traces to the UCITS issuer facts the universe audit found checkable: fund size, income policy and the wrapper's own registration
- `data/quality.py::run` traces to the eight stops and four warnings of this effort, each taken from the European UCITS universe audit's findings, plus the issuer facts the UCITS rules make checkable
- `data/quality.py::stop_dividend_blind` traces to the distribution-blindness stop of this effort: a price series whose adjusted close never diverges from its close is refused rather than patched
- `data/quality.py::stop_implausible_prices` traces to the eight stops and four warnings of this effort, each taken from the European UCITS universe audit's findings, plus the issuer facts the UCITS rules make checkable
- `data/quality.py::stop_late_start` traces to the eight stops and four warnings of this effort, each taken from the European UCITS universe audit's findings, plus the issuer facts the UCITS rules make checkable
- `data/quality.py::stop_malformed_dates` traces to the eight stops and four warnings of this effort, each taken from the European UCITS universe audit's findings, plus the issuer facts the UCITS rules make checkable
- `data/quality.py::stop_missing_bars` traces to the eight stops and four warnings of this effort, each taken from the European UCITS universe audit's findings, plus the issuer facts the UCITS rules make checkable
- `data/quality.py::stop_missing_columns` traces to the eight stops and four warnings of this effort, each taken from the European UCITS universe audit's findings, plus the issuer facts the UCITS rules make checkable
- `data/quality.py::stop_too_fresh` traces to the eight stops and four warnings of this effort, each taken from the European UCITS universe audit's findings, plus the issuer facts the UCITS rules make checkable
- `data/quality.py::warn_extra_distributions` traces to the eight stops and four warnings of this effort, each taken from the European UCITS universe audit's findings, plus the issuer facts the UCITS rules make checkable
- `data/quality.py::warn_factor_trim` traces to the eight stops and four warnings of this effort, each taken from the European UCITS universe audit's findings, plus the issuer facts the UCITS rules make checkable
- `data/quality.py::warn_thin_liquidity` traces to the eight stops and four warnings of this effort, each taken from the European UCITS universe audit's findings, plus the issuer facts the UCITS rules make checkable
- `data/quality.py::warn_tr_divergence` traces to the eight stops and four warnings of this effort, each taken from the European UCITS universe audit's findings, plus the issuer facts the UCITS rules make checkable
- `data/external.py::accrue_monthly` traces to the accrual rule the European Central Bank's published basis forces: its rate is annualised, so a month's return is the rate divided by 360 over that month's days rather than the rate compounded per period
- `data/external.py::eur_translate` traces to the translation leg of this effort: a sleeve's euro return is `(1 + local)(1 + translation) - 1`, so the currency line is a term rather than a residual
- `data/external.py::monthly_last` traces to the external legs: the ECB Data Portal's monthly series in the position it publishes them, the Fama/French archive's monthly returns, and the European Central Bank's daily/annual basis distinction with the /360 accrual it forces
- `data/external.py::monthly_returns` traces to the external legs: the ECB Data Portal's monthly series in the position it publishes them, the Fama/French archive's monthly returns, and the European Central Bank's daily/annual basis distinction with the /360 accrual it forces
- `data/external.py::parse_french_zip` traces to the Fama/French monthly archive and its two parser traps: the momentum column is WML, and the annual rows follow the monthly ones
- `data/external.py::read_ecb_csv` traces to the external legs: the ECB Data Portal's monthly series in the position it publishes them, the Fama/French archive's monthly returns, and the European Central Bank's daily/annual basis distinction with the /360 accrual it forces
- `data/external.py::splice_risk_free` traces to the risk-free splice of this effort: EONIA and the euro short-term rate joined at the euro short-term rate's first observation, whose overlap was measured at +8.5 bp on 579 days

## 2. Why it works this way, including what was rejected

_The module headers, verbatim: each records why the module is shaped the way it is, what was rejected, and the measurement that settled it. They are quoted here rather than restated, so the notebook cannot drift from the code._

**`data/coverage.py`**

The coverage report: what the frozen panel actually holds.

Run from the repository root:

    python3 -m portfolio_workbench.data.coverage
    python3 -m portfolio_workbench.data.coverage .data/panel/2026-09-13

It writes nothing and asserts nothing. Its job is to put the panel's shape, its
warnings and the cost of the joins in front of a reader before any result is quoted.

**`data/loader.py`**

Reading the frozen snapshot.

The loader is the only way into the panel: it verifies the manifest, parses each file
into the long table the rest of the package consumes, runs the quality gate, and hands
back the joined legs as one named document. Anything reading a snapshot file directly
bypasses the check, which is why the fetch script writes its manifest through
`manifest.describe` rather than by hand.

**The document is the seam, so its members are names rather than keys.** Every consumer used to
reach into a bare dict - eighty-nine key reads across thirteen modules, with the same three keys
re-typed into `panel.eur_excess_returns` at eleven of them - and a key renamed here surfaced as a
`KeyError` in whichever report ran first rather than as a failed check. The document also carries
the analytics frames derived from the legs (`returns`, `split`), because the loader is the one
place that holds every leg at once and the frame eleven call sites each rebuilt is a frame eleven
call sites can rebuild differently.

**The manifest is a document of its own, and is not namespaced here.** It mirrors the file the
snapshot ships, with the keys `data/manifest.py` measures and verifies, and it travels as the
mapping it is read as.

The long table is the data contract: one row per instrument-month, carrying
`period_month`, `available_from`, the unadjusted close, the feed's adjusted close and the
distribution. It is what another project would have to satisfy to consume this engine.

**`data/manifest.py`**

The snapshot manifest, and the check that the snapshot is the one it claims to be.

The panel is fetched once and then frozen; every result is keyed to the snapshot id it
was read from. The check is fail-closed in the strict sense: a checksum, row-count or
date-range mismatch raises instead of warning, because a panel that has silently
changed underneath a result makes the result uninterpretable, not merely stale.

**`data/panel.py`**

The as-of rule and the returned series.

A monthly bar is labelled with the month's first day while its close is the month's
last close. Nothing in the package may read such a bar before the first day of the
following month, and the only way to read the panel is through `as_of`, so the
embargo is a property of the access path rather than a rule each caller remembers.

**`data/quality.py`**

The quality gate: eight stops and four warnings, as rules rather than as surprises.

A stop means the panel is not the panel the results are keyed to, and it raises. A
warning means the number is usable but qualified, and it prints beside the number it
qualifies - never in a footnote, because a qualification a reader has to go looking for
is a qualification that gets dropped.

The eighth stop, a manifest mismatch, is enforced by `manifest.verify` before any frame
is built, since it is a property of the files rather than of the loaded table, and again
by `issuer_facts` for every priced line that reaches a rule reading the issuer facts: a
manifest that describes no issuer is a manifest fault, and reading the absence as "does
not distribute" is what would let the dividend-blind stop and the liquidity floor pass a
panel neither of them examined.

**`data/universe.py`**

The frozen universe, the mandate constants and the panel window.

Every value here was fixed before the build began, with one exception that is derived from the maps
above it rather than restated: `INSTRUMENT_GROUP` is the only place a ticker is joined to a budget
group, so a second copy cannot drift from the sleeve map and leave the budget aggregating a book it was
not declared against.
Insertion order is load-bearing. `TICKERS` is built from `SLEEVES`, and every weight
vector in the package is indexed by `TICKERS`, so reordering the sleeve map would
silently misalign weights against returns without raising anything.

**`data/external.py`**

External series: the central-bank risk-free chain, the FX leg and the factor archives.

Three traps live in this module and each cost an executed check to find, so they are
encoded as code with the reason beside them rather than left to the caller.

## 3. The data contract it consumes, and the as-of rule

One row per instrument-month, carrying `period_month` and `available_from`, plus the manifest that keys the whole snapshot. **A month's bar is available only from the first day of the month after the month it describes**, so a decision dated to a month can use that month's bar only if the decision is taken after its close. The price file's adjusted close is the total-return proxy, and a series whose adjusted close never diverges from its close is refused rather than patched, because a distribution-blind series would understate every return it carries. Euro returns are the local leg translated at the month's rate, less the overnight rate accrued over the month's days at the annualised quote divided by 360.

## 4. The worked example on small numbers, with the identity checked

The as-of rule is the whole of the look-ahead defence, so the worked example is that rule on three months: each bar becomes available the month after it, and a decision dated to March can see February's bar and not March's.

The cell below runs on numbers small enough to check by hand and asserts the identity, so a reader can see the arithmetic rather than take the module's word for it.

In [1]:
import pandas as pd

from portfolio_workbench.data import panel

months = pd.PeriodIndex(["2020-01", "2020-02", "2020-03"], freq="M")
frame = pd.DataFrame({"period_month": months, "available_from": panel.available_from(months)})

# Each bar is available from the first day of the following month.
assert list(frame["available_from"].dt.strftime("%Y-%m")) == ["2020-02", "2020-03", "2020-04"]

# A decision dated 2020-03-15 sees January's and February's bars, and not March's.
seen = panel.as_of(frame, pd.Timestamp("2020-03-15"))
assert list(seen["period_month"]) == list(months[:2])
print("bars visible on 2020-03-15:", [str(month) for month in seen["period_month"]])

bars visible on 2020-03-15: ['2020-01', '2020-02']


## 5. The real run: inputs, parameters, provenance block

The provenance block is printed first, then the parameters this module decides under, then the entry point's own report. The report is the module's output rather than a transcription of it, so a number quoted from a notebook is the number the module prints.

In [2]:
from portfolio_workbench.data import loader, universe

document = loader.load_panel()
months = document.months
print(f"snapshot {document.snapshot_id}, taken as of {document.as_of}")
print(f"panel {len(months)} months {months.min()}..{months.max()} across {len(universe.TICKERS)} sleeves")
print("manifest fields: " + ", ".join(sorted(document.manifest)))

from portfolio_workbench.data import panel, universe

print(f"sleeves {len(universe.TICKERS)}: {', '.join(universe.TICKERS)}")
print(f"panel window {universe.WINDOW_START}..{universe.WINDOW_END}, as-of taken on the manifest's own date")

snapshot 2026-09-13, taken as of 2026-09-13T07:56:28+00:00
panel 191 months 2010-09..2026-07 across 11 sleeves
manifest fields: created, excluded, files, instruments, snapshot_id, window
sleeves 11: IWDA.AS, IMEU.AS, XACT-NORDEN.ST, IBGL.AS, IEGE.AS, IEAC.AS, IHYG.L, IBCI.AS, 4GLD.DE, IWDP.AS, XEON.DE
panel window 2010-09..2026-08, as-of taken on the manifest's own date


In [3]:
import subprocess
import sys

finished = subprocess.run(
    [sys.executable, "-m", "portfolio_workbench.data.coverage"], capture_output=True, text=True, cwd="."
)
print(finished.stdout)
assert finished.returncode == 0, finished.stderr

[data] snapshot 2026-09-13, taken 2026-09-13T07:56:28+00:00
[data] 2112 rows over 11 instruments, 192 months of history
[data] joined panel 2010-09 → 2026-07 = 191 months × 11 sleeves
[data] coverage per instrument:
    4GLD.DE          2010-09 → 2026-08  192 months  distributions 0
    IBCI.AS          2010-09 → 2026-08  192 months  distributions 0
    IBGL.AS          2010-09 → 2026-08  192 months  distributions 27
    IEAC.AS          2010-09 → 2026-08  192 months  distributions 25
    IEGE.AS          2010-09 → 2026-08  192 months  distributions 10
    IHYG.L           2010-09 → 2026-08  192 months  distributions 28
    IMEU.AS          2010-09 → 2026-08  192 months  distributions 55
    IWDA.AS          2010-09 → 2026-08  192 months  distributions 0
    IWDP.AS          2010-09 → 2026-08  192 months  distributions 51
    XACT-NORDEN.ST   2010-09 → 2026-08  192 months  distributions 2
    XEON.DE          2010-09 → 2026-08  192 months  distributions 0
[data] FX series: ['EURSEK=X',

## 6. Results, and how to read them, including the resolution limit and what a reader must not conclude

The panel's shape is the resolution limit of every number downstream: 191 joined months from 2010-09 to 2026-07, of which 131 are traded out of sample, and eleven sleeves of which four are priced from the first month of the panel. A reader must not read a coverage gap as a verdict on the sleeve: the emerging-market sleeve is absent by decision rather than by data failure, and a sleeve that starts late is reported with its start rather than filled backwards. Nor is a thin liquidity warning a statement about the sleeve's tradability at size, which this panel cannot observe.

## 7. What this module does not establish

Nothing here establishes anything about instruments outside the eleven sleeves, about intra-month timing, or about execution: the panel is monthly and the prices are closing marks. It does not establish that the sleeves are the right exposures for the mandate, and the coverage report is a description of what was read rather than a judgement on the source.